dc/s2n-quic-dc/src/fixed_map.rs

```diff
@@ -0,0 +1,170 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+//! A fixed-allocation concurrent HashMap.
+//!
+//! This implements a concurrent map backed by a fixed-size allocation created at construction
+//! time, with a fixed memory footprint. The expectation is that all storage is inline (to the
+//! extent possible) reducing the likelihood.
+
+use core::{
+    hash::Hash,
+    sync::atomic::{AtomicU8, Ordering},
+};
+use parking_lot::{MappedRwLockReadGuard, RwLock, RwLockReadGuard, RwLockUpgradableReadGuard};
+use std::{collections::hash_map::RandomState, hash::BuildHasher};
+
+pub struct Map<K, V, S = RandomState> {
+    slots: Box<[Slot<K, V>]>,
+    hash_builder: S,
+}
+
+impl<K, V, S> Map<K, V, S>
+where
+    K: Hash + Eq,
+    S: BuildHasher,
+{
+    pub fn with_capacity(entries: usize, hasher: S) -> Self {
+        let map = Map {
+            slots: (0..std::cmp::min(1, (entries + SLOT_CAPACITY) / SLOT_CAPACITY))
+                .map(|_| Slot::new())
+                .collect::<Vec<_>>()
+                .into_boxed_slice(),
+            hash_builder: hasher,
+        };
+        assert!(map.slots.len().is_power_of_two());
+        assert!(u32::try_from(map.slots.len()).is_ok());
+        map
+    }
+
+    pub fn clear(&self) {
+        for slot in self.slots.iter() {
+            slot.clear();
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.slots.iter().map(|s| s.len()).sum()
+    }
+
+    // can't lend references to values outside of a lock, so Iterator interface doesn't work
+    #[allow(unused)]
+    pub fn iter(&self, mut f: impl FnMut(&K, &V)) {
+        for slot in self.slots.iter() {
+            // this feels more readable than flatten
+            #[allow(clippy::manual_flatten)]
+            for entry in slot.values.read().iter() {
+                if let Some(v) = entry {
+                    f(&v.0, &v.1);
+                }
+            }
+        }
+    }
+
+    pub fn retain(&self, mut f: impl FnMut(&K, &V) -> bool) {
+        for slot in self.slots.iter() {
+            // this feels more readable than flatten
+            #[allow(clippy::manual_flatten)]
+            for entry in slot.values.write().iter_mut() {
+                if let Some(v) = entry {
+                    if !f(&v.0, &v.1) {
+                        *entry = None;
+                    }
+                }
+            }
+        }
+    }
+
+    fn slot_by_hash(&self, key: &K) -> &Slot<K, V> {
+        let hash = self.hash_builder.hash_one(&key);
```

Comment:
cliipy issue: https://github.com/aws/s2n-quic/actions/runs/11079153316/job/30787689537?pr=2337#step:5:1064
```suggestion
        let hash = self.hash_builder.hash_one(key);
```